### PR TITLE
fix `Math.pow_body` for huge powers

### LINKED ISSUE #53881: BUG: `^(::Float64, ::Union{Int, Float64})` incorrect for very large negative exponents
The floating point power function fails for a border case of integer argument:

```
julia> 2.0 ^ typemin(Int) # should be 0.0
0.5

julia> (-1.0) ^ typemin(Int) # should be 1.0
-1.0
```

The bug seems to be here: [Base.Math.pow_body(::Float64, ::Integer)](https://github.com/JuliaLang/julia/blob/e07c0f1ddbfc89ad1ac4dda7246d8ed5d0d57c19/base/math.jl#L1326)

### PR DESCRIPTION
**Pull request by KlausC**
Fix `Math.pow_body(a, n)` for huge values of `n`.

Fixes #53881

### CODE REVIEW

File: base/math.jl [hunk 1]
```diff
@@ -1316,17 +1316,34 @@ end
 
 @assume_effects :terminates_locally @noinline function pow_body(x::Float64, n::Integer)
     y = 1.0
-    xnlo = ynlo = 0.0
+    xnlo = ynlo = err = 0.0
+    if iszero(n)
+        return y
```
Comment (oscardssmith) [inline base/math.jl:1321]: this shouldn't be necessary (see line 1313).

File: base/math.jl [hunk 2]
```diff
@@ -1316,17 +1316,34 @@ end
...
+    elseif n == 1
```
Comment (oscardssmith) [inline base/math.jl:1322]: this also shouldn't be necessary.

File: base/math.jl [hunk 3]
```diff
@@ -1316,17 +1316,34 @@ end
...
+        return x
+    elseif n == -1
```
Comment (oscardssmith) [inline base/math.jl:1324]: neither should this.

File: base/math.jl [hunk 4]
```diff
@@ -1316,17 +1316,34 @@ end
...
+        return inv(x)
+    end
+    x0 = x
     n == 3 && return x*x*x # keep compatibility with literal_pow
+    if isodd(n)
+        y = n < 0 ? inv(x) : x
```
Comment (oscardssmith) [inline base/math.jl:1330]: this seems broken. Where's the compensation?

**Comment by oscardssmith** [comment]
Is my understanding correct that the theory behind this is that it moves the first iteration out of the loop? It's somewhat unclear to me why this makes a difference.

**Comment by KlausC** [comment]
> Is my understanding correct that the theory behind this is that it moves the first iteration out of the loop? It's somewhat unclear to me why this makes a difference.

Yes, I took the first execution out of the loop, because I wanted to half the size of `n`, so `n = -n` does not break.
As a side-effect, the inverse of `x` was not required, but of `x * x`, which made the difference for `x = prevfloat(1.0)`.

File: base/math.jl [hunk 5]
```diff
@@ -1327,7 +1327,12 @@ end
     x0 = x
     n == 3 && return x*x*x # keep compatibility with literal_pow
     if isodd(n)
-        y = n < 0 ? inv(x) : x
+        if n < 0
+            y = inv(x)
+            ynlo = -fma(x, y, -1.0) * y
```
Comment (oscardssmith) [inline base/math.jl:1332]: I would also rewrite this to `ynlo = fma(-x, y, 1.0)* y`. I'm not sure why I did it with the double negation previously.

File: base/math.jl [hunk 6]
```diff
@@ -1341,7 +1341,7 @@ end
         n == -2 && return rx #keep compatibility with literal_pow
         !isfinite(rx) && return isodd(n) ? copysign(rx, x0) : rx
         if isfinite(x)
-            xnlo = -(xnlo * rx + fma(x, rx, -1.0)) * rx
+            xnlo = (fma(-x, rx, 1.0) - xnlo * rx) * rx
```
Comment (oscardssmith) [inline base/math.jl:1344]: make the `-xnlo*rx` a muladd.

**Comment by oscardssmith** [comment]
Overall this looks good. I do want to do a little more testing with random giant comments to make sure this fully fixes the issues.

**Review by KlausC — COMMENTED** [review]
The previous test run failed for two 32-bit AMD cpus. Reverted `muladd` to `fma` to check out, if that's the reason.

That was indeed the reason. So on these architectures, `muladd` behaves worse compared to and `fma`!!!

**Comment by KlausC** [comment]
The `pow_body` method was face-lifted a little bit. Now it has a accuracy of a few ulps in the complete definition area with no performance regressions. Please see the reported results:

```
function ulps(x, n, p=^)
           a = p(x, n)
           b = big(x) ^ n
           u = abs(a - Float64(b)) / eps(a)
           return isfinite(u) ? Int128(ceil(u)) : -1
 end

function stat_ulps(m)
           M1, M2 = log(floatmin(Float64)), log(floatmax(Float64))
           l1, l2 = sqrt(log2(3)), sqrt(log2(log2(floatmax()) / log2(nextfloat(1.0))))
           D = Dict{Integer,Any}()
           try
             for i = 1:m
               ab = rand() * (M2 - M1) + M1
               lb = (rand() * (l2 - l1) + l1) ^ 2
               b = Int64(floor(2.0 ^ lb)) * rand([-1,1])
               ab /= b
               s = sign(ab)
               la = prevfloat(abs(ab))
               c = expm1(la)
               a = (c - eps(c + 1) / 2) + 1
               a = s < 0 ? inv(a) : a
               if !isfinite(a)
                   println("ab=$ab lb=$lb s=$s b=$b la=$la c=$c a=$a")
               end
               u = ulps(a, b)
               R = get!(D, u) do; [] end
               t = (b, a)
               push!(R, t)
             end
           finally
               ;
           end
           for k in keys(D)
               D[k] = sort(collect(Set(D[k])); by = t-> abs(t[1]) + log2(abs(t[2])+eps()) * eps())
           end
           D
end

 function overview(D::Dict{<:Integer,<:Any})
         s0 = s1 = s2 = 0.0
         m = NaN
         ky = sort!(collect(keys(D)))
         n = sum(length(D[k]) for k in ky); println(n)
          for k in ky
                 l = float(length(D[k]))
                 if s0 + l > n / 2 && isnan(m)
                      m = k - 0.5 + (n/2 - s0) / l
                 end
                 s0 += l
                 s1 += l * k
                 s2 += l * k * k
                  println("$k => $l")
           end
           u = round(s1 / s0; digits = 3)
            v = round(sqrt( s2 / s0 - u * u); digits=3)
            u, m, v = round.((u, m, v); digits=3)
            println("total: n: $s0 - median: $m - mean: $u ± $v ulps")
end

julia> D = @time stat_ulps(10^6);
  6.709275 seconds (9.64 M allocations: 471.625 MiB, 8.45% gc time, 2.49% compilation time)

julia> overview(D)
1000000
0 => 768967.0
1 => 230895.0
2 => 77.0
3 => 39.0
4 => 9.0
5 => 8.0
6 => 4.0
7 => 1.0
total: n: 1.0e6 - median: 0.15 - mean: 0.231 ± 0.423 ulps

julia> @benchmark x^n setup= n, x = rand(D[0])
BenchmarkTools.Trial: 10000 samples with 997 evaluations.
 Range (min … max):  18.704 ns …  6.537 μs  ┊ GC (min … max): 0.00% … 98.42%
 Time  (median):     69.834 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   80.907 ns ± 84.292 ns  ┊ GC (mean ± σ):  1.07% ±  1.38%

   ▂█▂                                                         
  ▃███▇▇▇▆▅▄▅▅▅▅▅▅▅▃▄▄▅▅▄▅▅▄▃▄▃▄▄▄▄▄▃▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▃
  18.7 ns         Histogram: frequency by time         253 ns <

 Memory estimate: 16 bytes, allocs estimate: 1.

julia> @benchmark x^n setup= n, x = rand(D[1])
BenchmarkTools.Trial: 10000 samples with 997 evaluations.
 Range (min … max):  18.748 ns …  5.083 μs  ┊ GC (min … max): 0.00% … 98.91%
 Time  (median):     72.091 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   83.205 ns ± 74.017 ns  ┊ GC (mean ± σ):  0.81% ±  1.31%

   ▅█▄▃▃▂▁                                                     
  ▄███████▇▅▇██▇▇▇▆▅▅▇▆▅▆▆▆▅▄▄▅▆▅▅▅▄▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▃
  18.7 ns         Histogram: frequency by time         265 ns <

 Memory estimate: 16 bytes, allocs estimate: 1.

julia> @benchmark x^n setup= n, x = rand(D[2])
BenchmarkTools.Trial: 10000 samples with 949 evaluations.
 Range (min … max):   96.888 ns …  5.754 μs  ┊ GC (min … max): 0.00% … 94.29%
 Time  (median):     280.251 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   276.585 ns ± 96.419 ns  ┊ GC (mean ± σ):  0.30% ±  1.29%

                                  █▁▂▃    ▁                     
  ▄▁▂▄▂▄▂▄▄▆▅▂▂▂▂▃▄▃▂▂▄▂▂▅▂▂▄▂▂▃▆▆████▂▅▃▇█▃▂▂▄▅▅▄▅▅▆▆▄▃▂▄▂▂▂▃ ▃
  96.9 ns         Histogram: frequency by time          413 ns <

 Memory estimate: 16 bytes, allocs estimate: 1.

julia> @benchmark x^n setup= n, x = rand(D[3])
BenchmarkTools.Trial: 10000 samples with 852 evaluations.
 Range (min … max):  143.079 ns …  6.419 μs  ┊ GC (min … max): 0.00% … 94.95%
 Time  (median):     304.870 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   311.604 ns ± 83.042 ns  ┊ GC (mean ± σ):  0.20% ±  0.95%

                       ▄▁▁█  ▅▇ ▂   ▁▂                          
  ▅▂▁▂▂▂▁▁▁▁▁▁▁▅▃▁▁▁▁▃▅████▇▇██▄█▅▇▆██▆▂▅▆▂▂▂▂▂▂▂▂▂▂▂▂▂▁▅▃▁▁▁▄ ▃
  143 ns          Histogram: frequency by time          490 ns <

 Memory estimate: 16 bytes, allocs estimate: 1.
```

**Comment by KlausC** [comment]
This is now ready to be reviewed / merged. Note that it is an essential improvement in accuracy (only 150 ppm more than 1 ULP error) for big exponents, while performance is maintained. (The current implementation has more than 34% in the same sampled argument area)

**Comment by KlausC** [comment]
The statistics are perfect now: 99.4 % of the samples now show 0 ULPS, 0.6% 1 ULPS error, no sample with more!

```
# Testing code:
using BenchmarkTools

function ulps(x, n, p=^)
           a = p(x, n)
           b = big(x) ^ n
           u = abs(a - Float64(b)) / eps(a)
           return isfinite(u) ? BigInt(ceil(u)) : -1
end

function randpower(; minex=3, maxex=typemax(Int64))
           maxex = min(maxex, log2(floatmax()) / log2(nextfloat(1.0)))
           minex = max(minex, 1)
           l1, l2 = sqrt(log2(minex)), sqrt(log2(maxex))
           ab = NaN
           while !isfinite(ab)
               ab = reinterpret(Float64, rand(UInt64))
           end
           s = sign(ab)
           ab = log(abs(ab))
           lb = (rand() * (l2 - l1) + l1) ^ 2
           b = Int64(floor(2.0 ^ lb)) * rand([-1, 1])
           ab /= b
           a = exp(ab) * s
           c = a^b
           if !isfinite(c)
               a *= s
               b = b ÷ 2
               c = a^b
           end
           return b, a
 end

function stat_ulps(nx::AbstractVector{<:Tuple{<:Integer,<:AbstractFloat}})
           D = Dict{Integer,Any}()
           try
               for (b, a) in nx
                   u = ulps(a, b)
                   R = get!(D, u) do; [] end;
                   t = (b, a)
                   push!(R, t)
               end
           finally
               ;
           end
           for k in keys(D)
               D[k] = sort(collect(Set(D[k])); by = t-> abs(t[1]) + log2(abs(t[2])+eps()) * eps())
           end
           return D
 end
function overview(D::Dict{<:Integer,<:Any}, maxulps=53)
                  s0 = s1 = s2 = 0.0
                  m = NaN
                  ky = sort!(collect(keys(D)))
                  n = sum(length(D[k]) for k in ky); println(n)
                  smax = 0
                  for k in ky
                      l = float(length(D[k]))
                      k > maxulps && (k = maxulps + 1)
                      if s0 + l > n / 2 && isnan(m)
                          m = k - 0.5 + (n/2 - s0) / l
                      end
                      s0 += l
                      s1 += l * k
                      s2 += l * k * k
                      if k <= maxulps
                          println("$k => $l")
                      else
                          smax += l
                      end
                  end
                  if smax > 0
                      println(">$maxulps => $smax")
                  end
                  u = round(s1 / s0; digits = 3)
                  v = round(sqrt( s2 / s0 - u * u); digits=3)
                  u, m, v = Float64.(round.((u, m, v); digits=3))
                  println("total: n: $s0 - median: $m - mean: $u ± $v ulps")
       end

nx = [randpower() for _ in 1:10^6];
D = stat_ulps(nx);
overview(D)

@benchmark [x^n for (n, x) in nx]
```

Output with the version of this PR:
```
julia> nx = [randpower() for _ in 1:10^6];

julia> D = stat_ulps(nx);

julia> overview(D)
1000000
0 => 994416.0
1 => 5584.0
total: n: 1.0e6 - median: 0.003 - mean: 0.006 ± 0.074 ulps

julia> @benchmark [x^n for (n, x) in nx]
BenchmarkTools.Trial: 46 samples with 1 evaluation.
 Range (min … max):  108.233 ms … 117.334 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     108.887 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):   109.416 ms ±   1.533 ms  ┊ GC (mean ± σ):  0.24% ± 0.44%

   █                                                             
  ▄█▆▁▁▁▇▅▃▃▁▁▁▁▃▇▃▁▁▃▁▃▃▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▃ ▁
  108 ms           Histogram: frequency by time          117 ms <

 Memory estimate: 7.63 MiB, allocs estimate: 4.
```

The output of the same commands on v10.2
```
julia> nx = [randpower() for _ in 1:10^6];

julia> D = stat_ulps(nx);

julia> overview(D, 10)
1000000
0 => 627937.0
1 => 18924.0
2 => 6065.0
3 => 4011.0
4 => 3136.0
5 => 2520.0
6 => 2157.0
7 => 1774.0
8 => 1626.0
9 => 1494.0
10 => 1303.0
>10 => 329053.0
total: n: 1.0e6 - median: 0.296 - mean: 3.753 ± 5.142 ulps

julia> @benchmark [x^n for (n, x) in nx]
BenchmarkTools.Trial: 53 samples with 1 evaluation.
 Range (min … max):  94.458 ms … 101.908 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     94.580 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):   95.161 ms ±   1.386 ms  ┊ GC (mean ± σ):  0.13% ± 0.44%

  ▃█▂                                                           
  ███▅▁▁▅▁▁▇▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█▅▇▁▁▁▁▁▅▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▅▁▁▅ ▁
  94.5 ms       Histogram: log(frequency) by time      98.8 ms <

 Memory estimate: 7.63 MiB, allocs estimate: 3.
```

**Comment by oscardssmith** [comment]
That looks great! It's also really nice to have a second person that actually understands this code well. Can you also post the benchmarks for `nx = rand(1:20, 10^6)`? I do think in the real world most uses are with pretty small powers so it would be good to make sure that those don't regress.

**Comment by KlausC** [comment]
In the realm of the old implementation, exponents are less than 2^25, the results are comparable both in accuracy and performance:

The benchmark evaluates 10^6 powers and requires 59 or 51 ns mean value.

New:
```
julia> nx = [randpower(maxex = 2^25) for _ in 1:10^6];

julia> D = stat_ulps(nx);

julia> overview(D)
1000000
0 => 994076.0
1 => 5924.0
total: n: 1.0e6 - median: 0.003 - mean: 0.006 ± 0.077 ulps

julia> @benchmark [x^n for (n, x) in nx]
BenchmarkTools.Trial: 84 samples with 1 evaluation.
 Range (min … max):  59.079 ms … 67.466 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     59.197 ms              ┊ GC (median):    0.00%
 Time  (mean ± σ):   59.655 ms ±  1.073 ms  ┊ GC (mean ± σ):  0.46% ± 0.68%

  ▅█▂         ▂▄▂▁                                             
  ███▅▁▁▁▁▁▁▁▁████▅▅▅▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▅▁▁▁▇▅▁▁▅▁▁▁▁▁▁▁▁▁▁▁▁▅ ▁
  59.1 ms      Histogram: log(frequency) by time      62.3 ms <

 Memory estimate: 7.63 MiB, allocs estimate: 4.
```

For comparison the old implementation:
```
julia> nx = [randpower(maxex = 2^25) for _ in 1:10^6];

julia> D = stat_ulps(nx);

julia> overview(D, 10)
1000000
0 => 993824.0
1 => 6176.0
total: n: 1.0e6 - median: 0.003 - mean: 0.006 ± 0.078 ulps

julia> @benchmark [x^n for (n, x) in nx]
BenchmarkTools.Trial: 97 samples with 1 evaluation.
 Range (min … max):  51.343 ms … 59.751 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     51.445 ms              ┊ GC (median):    0.00%
 Time  (mean ± σ):   51.942 ms ±  1.381 ms  ┊ GC (mean ± σ):  0.11% ± 0.36%

  █▃             ▁                                             
  ██▄▁▄█▄▁▁▁▁▁▁▁▁█▆▁▁▁▆▄▁▁▁▁▁▁▁▁▁▁▁▁▁▁▄▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▄ ▁
  51.3 ms      Histogram: log(frequency) by time      59.2 ms <

 Memory estimate: 7.63 MiB, allocs estimate: 3.
```

Baseline for the benchmark:
```
julia> @benchmark [x*n for (n, x) in nx]
BenchmarkTools.Trial: 3076 samples with 1 evaluation.
 Range (min … max):  1.493 ms …   3.959 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     1.543 ms               ┊ GC (median):    0.00%
 Time  (mean ± σ):   1.620 ms ± 214.669 μs  ┊ GC (mean ± σ):  2.72% ± 6.65%

  ▂▆██▇▆▄▃▂▂▂▂▂                               ▁▄▃▂▁           ▁
  ██████████████▇▇██▇▇▆▇▆▅▄▆▆▇▅▅▄▆▅▅▅▆▄▄▄▅▄▆▆▅█████▇▇▇▅▅▅▄▁▄▅ █
  1.49 ms      Histogram: log(frequency) by time      2.22 ms <

 Memory estimate: 7.63 MiB, allocs estimate: 3.
```

**Comment by oscardssmith** [comment]
I also wonder whether it would be worth making it so that big powers go to the regular Float64^Float64 path. My bench-marking suggests that around `10000` the Float64 path is faster.

**Comment by KlausC** [comment]
> Can you also post the benchmarks for `nx = rand(1:20, 10^6)`

Do you mean restrict the exponents to -20:20 or to -10^6:10^6 ?

**Comment by KlausC** [comment]
> whether it would be worth making it so that big powers go to the regular Float64^Float64 path.

I will check, how performance verse accuracy works out.

**Comment by KlausC** [comment]
> I will check, how performance verse accuracy works out.

@oscardssmith yes, that would be a better way. I wasn't aware that we have a double precision implementation of log/exp behind the scenes. That has high accuracy for all integer exponents which can be represented as `Float64` and superior performance figures for exponents maybe `-10000 < n < 10000` (Threshold to be evaluated). For integers, which remain, (`Int64(Float64(n)) != n`, a special handling is required.

This is a separate PR, though. If that will be done, this PR is required not anymore.

Nevertheless it was a challenging task for me to get out the last ULP of the algorithm, having much fun!
Essential algorithmic imprevement was the merging of the hi and lo precision parts of the variables more or less frequently, before the lo part exploded in size. For the bases less than 1, with negative exponents, special handling was needed. That all came with a price of 20-30% performance reduction, compared with the existing implementation.

Note, that there is still issue #53881.

**Review by oscardssmith — APPROVED** [review]

**Comment by KlausC** [comment]
This PR is superseded with #53967 and closed.

**Pull request closed without merging**